include/openmc/tallies/tally_filter.h

```diff
@@ -0,0 +1,88 @@
+#ifndef OPENMC_TALLY_FILTER_H
+#define OPENMC_TALLY_FILTER_H
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "openmc/hdf5_interface.h"
+#include "openmc/particle.h"
+#include "openmc/xml_interface.h"
+
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+extern "C" int32_t n_filters;
+
+class TallyFilterMatch;
+extern std::vector<TallyFilterMatch> filter_matches;
+#pragma omp threadprivate(filter_matches)
+
+class TallyFilter;
+extern std::vector<TallyFilter*> tally_filters;
+
+//==============================================================================
+//! Stores bins and weights for filtered tally events.
+//==============================================================================
+
+class TallyFilterMatch
+{
+public:
+  //int i_bin_;
+  std::vector<int> bins_;
+  std::vector<double> weights_;
+  //bool bins_present_;
+};
+
+//==============================================================================
+//! Modifies tally score events.
+//==============================================================================
+
+class TallyFilter
+{
+public:
+  virtual std::string type() const = 0;
+
+  virtual ~TallyFilter() = 0;
+
+  //! Uses an XML input to fill the filter's data fields.
+  virtual void from_xml(pugi::xml_node node) = 0;
```

Comment:
It might be a little awkward with the interactions between the C-API and Fortran, but I'm not sure yet.  I'll get back to you.

Reply:
I looked into this a little more.  It's possible, but I don't think it would make the code cleaner at the moment due to Fortran interop and C-API concerns.  I think we should revisit this idea later when we've fully removed tally filters from Fortran.

Reply:
I suspected that might be the case. Thanks for looking into it!